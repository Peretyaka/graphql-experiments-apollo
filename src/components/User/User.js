import React from 'react'
import PropTypes from 'prop-types'
import Card from '../ui/Card'
import {Button, ButtonPrimary} from '../ui/Button'
import Avatar from '../Avatar'
import injectSheet from '../../utils/jss'

/**
 * @param {number} gender
 */
const getGenderTranslation = (gender) => {
  if (gender === 1) return 'Male'
  return 'Female'
}

const styles = {
  user: {
    padding: 20,
    display: 'flex',
    alignItems: 'flex-start',
  },
  avatar: {
    flexShrink: 1,
    flexGrow: 0,
    marginRight: 20,
  },
  content: {
    flexGrow: 1,
  },
  row: {
    marginBottom: 10,
  },
  screenname: {
    composes: '$row',
    fontWeight: 'bold',
  },
  inline: {
    display: 'inline-block',
    verticalAlign: 'middle',
    '&::before': {
      content: '","',
      display: 'inline',
      paddingRight: 5,
    },
    '&:first-child::before': {
      display: 'none',
    },
  },
  actions: {
    marginTop: 20,
    display: 'flex',
    alignItems: 'center',
  },
  action: {
    marginRight: 20,
  }
}

const User = ({classes, screenname, avatar, gender, city}) => (
  <Card className={classes.user}>
    <div className={classes.avatar}>
      <Avatar src={avatar} medium round />
    </div>
    <div className={classes.content}>
      <div className={classes.screenname}>{screenname}</div>
      <div>
        {gender && <div className={classes.inline}>{getGenderTranslation(gender)}</div>}
        {city && <div className={classes.inline}>from {city}</div>}
      </div>
      <div className={classes.actions}>
        <div className={classes.action}>
          <ButtonPrimary>Like</ButtonPrimary>
        </div>
        <div className={classes.action}>
          <Button>Add to friends</Button>
        </div>
      </div>
    </div>
  </Card>
)

User.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  avatar: PropTypes.string.isRequired,
  screenname: PropTypes.string.isRequired,
  gender: PropTypes.oneOf([1, 2]), // 1 - male, 2 - female
  city: PropTypes.string,
}

User.defaultProps = {
  gender: 0,
  city: null,
}

export default injectSheet(styles)(User)
