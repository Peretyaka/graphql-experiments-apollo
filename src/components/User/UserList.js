import React, {Children, PureComponent} from 'react'
import PropTypes from 'prop-types'
import injectSheet from '../../utils/jss'
import Loader from '../ui/Loader'
import {ButtonPrimary} from '../ui/Button'

const styles = {
  list: {
    textAlign: 'center',
    minHeight: 200,
  },
  action: {
    marginTop: 20,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  item: {
    display: 'inline-block',
    textAlign: 'left',
    verticalAlign: 'top',
    width: 385,
    margin: [0, 10, 20],
  },
}

class UserList extends PureComponent {
  static propTypes = {
    classes: PropTypes.objectOf(PropTypes.string).isRequired,
    children: PropTypes.node,
    loading: PropTypes.bool,
    onLoadMoreClick: PropTypes.func,
  }

  constructor(props) {
    super(props)

    this.defaultProps = {
      children: null,
      loading: false,
      onLoadMoreClick: null,
    }

    this.handleLoadMoreClick = this.handleLoadMoreClick.bind(this)
  }

  /**
   * Use as proxy method for passing updated count of rendered users
   */
  handleLoadMoreClick() {
    this.props.onLoadMoreClick(this.props.children.length)
  }

  render() {
    const {children, classes, loading, onLoadMoreClick} = this.props
    const showAction = !loading && onLoadMoreClick

    return (
      <Loader transparent active={loading}>
        <div className={classes.list}>
          {Children.map(children, child => (
            <div className={classes.item}>{child}</div>
          ))}
        </div>
        {showAction && (
          <div className={classes.action}>
            <ButtonPrimary
              onClick={this.handleLoadMoreClick}
            >
              Load more
            </ButtonPrimary>
          </div>
        )}
      </Loader>
    )
  }
}

export default injectSheet(styles)(UserList)
