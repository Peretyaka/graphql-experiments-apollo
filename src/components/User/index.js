export UserList from './UserList'
export default from './User'
