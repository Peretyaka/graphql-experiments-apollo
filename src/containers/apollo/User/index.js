export {default as UserWithData} from './UserWithData'
export {default as UserListWithData} from './UserListWithData'