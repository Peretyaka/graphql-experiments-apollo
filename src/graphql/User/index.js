export {default as queries} from './queries'
export {default as fragments} from './fragments'
