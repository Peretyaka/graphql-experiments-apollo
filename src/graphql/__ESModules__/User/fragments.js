import gql from 'graphql-tag'

export const DEFAULT_USER_FRAGMENT = gql`
  fragment defaultFields on User {
    id
    screenname
    avatar
    gender
    geo {
      city
      country
    }
  }
`
