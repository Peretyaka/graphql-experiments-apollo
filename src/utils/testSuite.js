import React from 'react'
import {create as createJss} from 'jss'
import preset from 'jss-preset-default'
import {JssProvider, ThemeProvider} from 'react-jss'
import Enzyme, {
  render as enzymeRender,
  shallow as enzymeShallow,
  mount as enzymeMount
} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import theme from '../theme'

// Setup enzyme adapter for current version of React
Enzyme.configure({
  adapter: new Adapter()
})

/**
 * Create testing jss instance
 * NOTE: For simplifying testing. Just return original classname
 */
const jss = createJss({
  createGenerateClassName: () => rule => rule.key,
  ...preset(),
})

/**
 * IMPORTANT: Find a solution to not pass theme provider
 * inside testing suites. Because any change of theme WILL break
 * all snapshots
 */
const renderComponentWithJss = children => (
  <ThemeProvider theme={theme}>
    <JssProvider jss={jss}>
      {children}
    </JssProvider>
  </ThemeProvider>
)

export const shallow = children => enzymeShallow(renderComponentWithJss(children))
export const render = children => enzymeRender(renderComponentWithJss(children))
export const mount = children => enzymeMount(renderComponentWithJss(children))
