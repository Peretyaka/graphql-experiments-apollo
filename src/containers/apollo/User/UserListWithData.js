import React from 'react'
import {Query} from 'react-apollo'
import gql from 'graphql-tag'
import {User, UserList} from '../../../components/User'

const ITEMS_PER_LOAD = 5

const QUERY = gql`
  query OffsetUsers($offset: Int!, $limit: Int!) {
    offsetUsers(offset: $offset, limit: $limit) {
      limitReached
      users {
        id
        avatar
        screenname
        gender
        geo {
          city
        }
      }
    }
  }
`

const UserListWithData = () => (
  <Query
    query={QUERY}
    variables={{
      offset: 0,
      limit: ITEMS_PER_LOAD
    }}
  >
    {({data, loading, fetchMore}) => {
      // If no data - just avoid rendering
      if (!data) return <div />

      return (
        <UserList
          loading={loading}
          limitReached={data.offsetUsers.limitReached}
          onLoadMoreClick={(offset) => {
            /**
             * We don't place handler in separated functon
             * due to need to pass 'fetchMore' function
             * as param to handler.
             */
            fetchMore({
              variables: {
                offset,
                limit: ITEMS_PER_LOAD,
              },
              updateQuery: (prevResult, {fetchMoreResult}) => {
                if (!fetchMoreResult) return prevResult
                /**
                 * Apollo doesn't do work for you in case of updating query.
                 * We need to manually extend previous data with new data
                 * to make new object that will be returned as new data
                 */
                return Object.assign({}, prevResult, {
                  offsetUsers: Object.assign({}, prevResult.offsetUsers, {
                    users: [
                      ...prevResult.offsetUsers.users,
                      ...fetchMoreResult.offsetUsers.users
                    ],
                    limitReached: fetchMoreResult.offsetUsers.limitReached
                  })
                })
              }
            })
          }}
        >
          {data.offsetUsers.users.map(({id, avatar, screenname, gender, geo: {city}}) => (
            <User
              key={id}
              avatar={avatar}
              screenname={screenname}
              gender={gender}
              city={city}
            />
          ))}
        </UserList>
      )
    }}
  </Query>
)

export default UserListWithData
