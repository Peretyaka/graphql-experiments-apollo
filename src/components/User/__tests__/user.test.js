import React from 'react'
import {shallow} from '../../../utils/testSuite'
import User from '../User'

describe('User', () => {
  it('Should render correctly', () => {
    const userList = shallow(<User />)
    expect(userList).toMatchSnapshot()
  })

  // TODO: Write more tests
})
