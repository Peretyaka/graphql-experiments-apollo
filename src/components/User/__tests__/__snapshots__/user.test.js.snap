// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ThemeProvider
    theme={
      Object {
        "brand": "#ff5959",
        "button": Object {
          "default": Object {
            "background": "#ebebeb",
            "color": "rgba(0,0,0,0.7)",
            "hoverBackground": "#e0e0e0",
            "icon": "rgba(0,0,0,0.5)",
          },
          "inverse": Object {
            "background": "#404040",
            "color": "rgba(255,255,255,0.9)",
            "hoverBackground": "#4d4d4d",
            "icon": "rgba(255,255,255,0.9)",
          },
          "primary": Object {
            "background": "#ff5959",
            "color": "rgba(255,255,255,0.9)",
            "hoverBackground": "#ff4545",
            "icon": "rgba(255,255,255,0.9)",
          },
        },
        "common": Object {
          "border": "#e6e6e6",
          "card": "#fff",
          "cardInverse": "#333",
          "header": "#141414",
          "page": "#f5f5f5",
          "radius": 3,
        },
        "danger": "#f1462f",
        "icon": Object {
          "default": "rgba(0,0,0,0.5)",
          "inverse": "rgba(255,255,255,0.9)",
        },
        "input": Object {
          "border": "rgba(0,0,0,0.15)",
          "label": "rgba(0,0,0,0.5)",
        },
        "loader": Object {
          "background": "rgba(0,0,0,0.2)",
          "foreground": "rgba(0,0,0,0.6)",
          "inverseBackground": "rgba(255,255,255,0.1)",
          "inverseForeground": "rgba(255,255,255,0.9)",
        },
        "shadow": Object {
          "level1": "0 0 2px 0 rgba(0,0,0,0.10), 0 2px 2px 0 rgba(0,0,0,0.20)",
          "level2": "0 0 4px 0 rgba(0,0,0,0.10), 0 4px 4px 0 rgba(0,0,0,0.20)",
          "level3": "0 0 8px 0 rgba(0,0,0,0.10), 0 8px 8px 0 rgba(0,0,0,0.20)",
          "level4": "0 0 16px 0 rgba(0,0,0,0.10), 0 16px 16px 0 rgba(0,0,0,0.20)",
          "level5": "0 0 24px 0 rgba(0,0,0,0.10), 0 24px 24px 0 rgba(0,0,0,0.20)",
        },
        "success": "#69b32d",
        "text": Object {
          "default": "rgba(0,0,0,0.85)",
          "highlight": "#ff5959",
          "inverse": "rgba(255,255,255,0.9)",
          "muted": "rgba(0,0,0,0.5)",
        },
        "typography": Object {
          "fontFamily": "Helvetica, Arial, sans-serif",
          "fontSize": 14,
          "lineHeight": 1.2,
        },
      }
    }
  >
    <JssProvider
      jss={
        Jss {
          "generateClassName": [Function],
          "options": Object {
            "Renderer": [Function],
            "createGenerateClassName": [Function],
            "plugins": Array [
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
                "onProcessRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
                "onProcessStyle": [Function],
                "onUpdate": [Function],
              },
              Object {
                "onProcessRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
                "onProcessRule": [Function],
              },
              Object {
                "onProcessStyle": [Function],
              },
              Object {
                "onProcessStyle": [Function],
              },
              Object {
                "onProcessStyle": [Function],
              },
              Object {
                "onProcessStyle": [Function],
              },
              Object {
                "onChangeValue": [Function],
                "onProcessStyle": [Function],
              },
              Object {
                "onProcessStyle": [Function],
              },
              Object {
                "onChangeValue": [Function],
                "onProcessRule": [Function],
                "onProcessStyle": [Function],
              },
              Object {
                "onProcessStyle": [Function],
              },
            ],
          },
          "plugins": PluginsRegistry {
            "hooks": Object {
              "onChangeValue": Array [
                [Function],
                [Function],
              ],
              "onCreateRule": Array [
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
              ],
              "onProcessRule": Array [
                [Function],
                [Function],
                [Function],
                [Function],
              ],
              "onProcessSheet": Array [],
              "onProcessStyle": Array [
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
              ],
              "onUpdate": Array [
                [Function],
              ],
            },
          },
          "version": "9.3.3",
        }
      }
    >
      <Jss(User)
        actions={null}
        city={null}
        gender={0}
      />
    </JssProvider>
  </ThemeProvider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Jss(User)
        actions={null}
        city={null}
        gender={0}
      />,
      "jss": Jss {
        "generateClassName": [Function],
        "options": Object {
          "Renderer": [Function],
          "createGenerateClassName": [Function],
          "plugins": Array [
            Object {
              "onCreateRule": [Function],
            },
            Object {
              "onCreateRule": [Function],
            },
            Object {
              "onCreateRule": [Function],
            },
            Object {
              "onCreateRule": [Function],
            },
            Object {
              "onCreateRule": [Function],
            },
            Object {
              "onCreateRule": [Function],
            },
            Object {
              "onCreateRule": [Function],
            },
            Object {
              "onCreateRule": [Function],
            },
            Object {
              "onCreateRule": [Function],
            },
            Object {
              "onCreateRule": [Function],
              "onProcessRule": [Function],
            },
            Object {
              "onCreateRule": [Function],
              "onProcessStyle": [Function],
              "onUpdate": [Function],
            },
            Object {
              "onProcessRule": [Function],
            },
            Object {
              "onCreateRule": [Function],
              "onProcessRule": [Function],
            },
            Object {
              "onProcessStyle": [Function],
            },
            Object {
              "onProcessStyle": [Function],
            },
            Object {
              "onProcessStyle": [Function],
            },
            Object {
              "onProcessStyle": [Function],
            },
            Object {
              "onChangeValue": [Function],
              "onProcessStyle": [Function],
            },
            Object {
              "onProcessStyle": [Function],
            },
            Object {
              "onChangeValue": [Function],
              "onProcessRule": [Function],
              "onProcessStyle": [Function],
            },
            Object {
              "onProcessStyle": [Function],
            },
          ],
        },
        "plugins": PluginsRegistry {
          "hooks": Object {
            "onChangeValue": Array [
              [Function],
              [Function],
            ],
            "onCreateRule": Array [
              [Function],
              [Function],
              [Function],
              [Function],
              [Function],
              [Function],
              [Function],
              [Function],
              [Function],
              [Function],
              [Function],
              [Function],
            ],
            "onProcessRule": Array [
              [Function],
              [Function],
              [Function],
              [Function],
            ],
            "onProcessSheet": Array [],
            "onProcessStyle": Array [
              [Function],
              [Function],
              [Function],
              [Function],
              [Function],
              [Function],
              [Function],
              [Function],
              [Function],
            ],
            "onUpdate": Array [
              [Function],
            ],
          },
        },
        "version": "9.3.3",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "actions": null,
        "city": null,
        "gender": 0,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Jss(User)
          actions={null}
          city={null}
          gender={0}
        />,
        "jss": Jss {
          "generateClassName": [Function],
          "options": Object {
            "Renderer": [Function],
            "createGenerateClassName": [Function],
            "plugins": Array [
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
                "onProcessRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
                "onProcessStyle": [Function],
                "onUpdate": [Function],
              },
              Object {
                "onProcessRule": [Function],
              },
              Object {
                "onCreateRule": [Function],
                "onProcessRule": [Function],
              },
              Object {
                "onProcessStyle": [Function],
              },
              Object {
                "onProcessStyle": [Function],
              },
              Object {
                "onProcessStyle": [Function],
              },
              Object {
                "onProcessStyle": [Function],
              },
              Object {
                "onChangeValue": [Function],
                "onProcessStyle": [Function],
              },
              Object {
                "onProcessStyle": [Function],
              },
              Object {
                "onChangeValue": [Function],
                "onProcessRule": [Function],
                "onProcessStyle": [Function],
              },
              Object {
                "onProcessStyle": [Function],
              },
            ],
          },
          "plugins": PluginsRegistry {
            "hooks": Object {
              "onChangeValue": Array [
                [Function],
                [Function],
              ],
              "onCreateRule": Array [
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
              ],
              "onProcessRule": Array [
                [Function],
                [Function],
                [Function],
                [Function],
              ],
              "onProcessSheet": Array [],
              "onProcessStyle": Array [
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
                [Function],
              ],
              "onUpdate": Array [
                [Function],
              ],
            },
          },
          "version": "9.3.3",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "actions": null,
          "city": null,
          "gender": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
